src/boot/bootctl-install.c

```diff
@@ -569,6 +574,161 @@ static int install_entry_token(void) {
+                        return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to write variable attributes to bio");
+
+                if (BIO_write(bio, &timestamp, sizeof(timestamp)) < 0)
+                        return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to write timestamp to bio");
+
+                if (BIO_write(bio, &siglist, siglist->SignatureListSize) < 0)
+                        return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to write signature list to bio");
+
+                _cleanup_(PKCS7_freep) PKCS7 *p7 = NULL;
+                p7 = PKCS7_sign(certificate, private_key, NULL, bio, PKCS7_DETACHED|PKCS7_NOATTR|PKCS7_BINARY|PKCS7_NOSMIMECAP);
+                if (!p7)
+                        return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to calculate PKCS7 signature: %s",
+                                               ERR_error_string(ERR_get_error(), NULL));
+
+                _cleanup_free_ uint8_t *sig = NULL;
+                int sigsz = i2d_PKCS7_SIGNED(p7->d.sign, &sig);
+                if (sigsz < 0)
+                        return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to convert PKCS7 signature to DER: %s",
+                                               ERR_error_string(ERR_get_error(), NULL));
+
+                _cleanup_free_ EFI_VARIABLE_AUTHENTICATION_2 *auth = malloc(sizeof(EFI_VARIABLE_AUTHENTICATION_2) + sigsz);
+                if (!auth)
+                        return log_oom();
+
+                *auth = (EFI_VARIABLE_AUTHENTICATION_2) {
+                        .TimeStamp = timestamp,
+                        .AuthInfo = {
+                                .Hdr = {
+                                        .dwLength = sizeof(auth->AuthInfo) + sigsz,
+                                        .wRevision = 0x0200,
+                                        .wCertificateType = 0x0EF1,
+                                },
+                                .CertType = EFI_CERT_TYPE_PKCS7_GUID,
+                        }
+                };
+
+                memcpy(auth->AuthInfo.CertData, sig, sigsz);
+
+                _cleanup_free_ char *filename = strjoin(db, ".auth");
+                if (!filename)
+                        return log_oom();
+
+                _cleanup_close_ int fd = openat(keys_fd, filename, O_CREAT|O_EXCL|O_NOFOLLOW|O_NOCTTY|O_WRONLY|O_CLOEXEC, 0600);
+                if (fd < 0)
+                        return log_error_errno(fd, "Failed to open secure boot auto-enrollment file for writing: %m");
+
+                r = loop_write(fd, auth, sizeof(*auth) + sigsz);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to write authentication descriptor to secure boot auto-enrollment file: %m");
+
+                r = loop_write(fd, siglist, siglist->SignatureListSize);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to write signature list to secure boot auto-enrollment file: %m");
+
```

Comment:
maybe fsync() both `fd` and `keys_fd` first before you write out the success message below. after all this is a shitty vfat